chainer/computational_graph.py

```diff
@@ -0,0 +1,201 @@
+        """
+        self.edges = edges
+
+    def _to_dot(self):
+        """Returns graph in dot format
+
+        Returns:
+            string: graph in dot format
+        """
+
+        ret = "digraph graphname{"
+        for edge in self.edges:
+            head, tail = edge
+            assert (isinstance(head, variable.Variable)
+                    and isinstance(tail, function.Function)) or \
+                   (isinstance(head, function.Function)
+                    and isinstance(tail, variable.Variable))
+            head_node = DotNode(head)
+            tail_node = DotNode(tail)
+            ret += head_node.label
+            ret += tail_node.label
+            ret += "%s -> %s;" % (head_node.id_, tail_node.id_)
+        ret += "}"
+        return ret
+
+    def dump(self, format='dot'):
+        """Dump graph as a text
+
+        Args
+            format: specifies graph language to output
+
+        Returns
+            string: graph in specified format
+        """
+        if format == 'dot':
+            return self._to_dot()
+        else:
+            NotImplementedError('Currently, only dot format is supported')
+
+    def __len__(self):
+        return len(self.edges)
+
+    def __contains__(self, e):
+        return e in self.edges
+
+
+def build_computational_graph(outputs, remove_split=True):
+    """Walk back from outputs to get graph whose nodes are reachable from outputs
+
+    Args:
+        outputs: List of nodes. Each node is either :class:`Variable` object
+        or :class:`Function` object.
+        remove_split : Boolean. If it is `True`, :class:`Split` functions
+        and cloned variables they created are skipped from resulting graph.
+
+    Returns:
+        :class:`ComputationalGraph`: consists of nodes and edges that
+        are reachable from `outputs`, with edges reversed.
+
+        For example, suppose that computational graph is as follows
+        (diagram is same as the one in the document of :class:`Function`):
+
+                               |--- x'  <--- f'  <--- y
+           x <--- (splitter) <-+
+                               |--- x'' <--- f'' <--- z
+
+        Let `outputs = [y, z]`.
+        If `remove_split` is `False`, this method generates the graph
+        itself with direction of edges reversed. If `remove_split` is `True`,
+        splitter and x' is removed from the graph and x is directly
+        connected to f'. After reversing the edges, reulting graph will be
+
+                     |---> y
+           x ---> f -+
+                     |---> z
+
+        Next, let `outputs = [y]`. Note that `z`, `f''`, and `x''`
+        is not reachable from `y`. If `remove_split` is `False`,
+        we remove these unreachable nodes to get
+
+           x ---> (splitter) ---> x'  ---> f'  ---> y
+
+        If `remove_split` is `True`, we further remove splitter and `x'` to get
+
+           x ---> f  ---> y
+
+        See `tests/test_computational_graph.py` for details.
+
```

Comment:
One more point: users do not concern about the actual graph implementation in Chainer (i.e., users may not know that Chainer maintains backward-oriented graphs). When one says "computational graph", one usually imagines a forward-oriented graph. I think the document of this function can be written without touching the backward-oriented implementation, and then the document gets simpler.